### PR TITLE
merl-an benchmark

### PR DESCRIPTION
**Pull request by 3Rafal**
This is the MVP of #1633

### CODE REVIEW

**Comment by voodoos** [comment]
Is there a way to get the locations of the slow queries ?
Some are surprising like type-enclosing taking 174772ms 😅

**Comment by 3Rafal** [comment]
@voodoos , we don't have a nice ux that lets you do this. This feature would be very helpful though.

**Comment by pitag-ha** [comment]
> Is there a way to get the locations of the slow queries ?
Some are surprising like type-enclosing taking 174772ms sweat_smile

That's almost certainly a "non-reproducible latency spike". What I mean with that is that with `merl-an` we recurrently observe the following: if you run the exact same Merlin query with the same Merlin state on the same location several times, one time it takes more than 10x longer than the other times. That's only "reproducible" in the sense that it's recurrent on all kind of projects, all kind of files, all kind of Merlin queries etc. We haven't been able to find one concrete example where we can observe it more repeatedly than on other examples, though.

Those "latency spikes" are also the reason why the `current-bench` graphs are scaled so badly: Their scale ranges from min timing to max timing. Max timing is over 10x higher than it should be due to spikes. I've already done things in `merl-an` to detect spikes in the past, it's not hard. So I think we should improve `merl-an benchmark` to filter out spikes. Possibly, instead of filtering them out completely, we could separate them into a different `current-bench` graph to also benchmark the number of spikes that happen.

I've just added three points to the [Benchmark CI issue](https://github.com/ocaml/merlin/issues/1633):
- Filter out spikes (as just mentioned)
- Add cold-cache workflow to the benchmarks
- Improve the output UX
 
All of those are just suggestions. I think this is a great PoC already!

File: bench.Dockerfile [hunk 1]
```diff
@@ -7,3 +7,26 @@ RUN sudo apt-get -y install jq
...
+RUN sudo mkdir /projects
+RUN sudo chown opam /projects
+WORKDIR /projects
+
+# build irmin
+RUN git clone https://github.com/mirage/irmin.git
+WORKDIR /projects/irmin
+RUN git checkout 8da4d16e7cc8beddfc8a824feca325426bae08a9
+RUN sudo apt install -y gnuplot-x11 libgmp-dev pkg-config libffi-dev
+RUN opam install . --deps-only --with-test --no-checksums
```
Comment (pitag-ha) [inline bench.Dockerfile:29]: Btw, one more nitpick: We might want to lock the dependencies for more determinism.

Comment (3Rafal) [reply bench.Dockerfile:30]: done

**Review by voodoos — COMMENTED** [review]
I am bit surprised by the need to have lock-files for every package. Maybe a switch export would have been simpler ?

I used it in https://github.com/voodoos/merlin-occurrences-switch

It allows you to lock a complete switch state and then reinstalling everything at the exact same version.

Also were you able to filter out the outliers ? We should try to investigate them at some point but for now it would be nice to get well scaled graph and more reasonnable results...

**Comment by 3Rafal** [comment]
I've managed to produce a list of top 10 slowest queries:
```
Top 10 slowest queries:
jq -s "[ (map(.sample_id |= tostring)) + (map(.sample_id |= tostring)) | group_by(.sample_id)[] | select(length > 1) | add ] | sort_by( .responses[0].timing.clock) | reverse | .[:10]" merl-an_bench/query_responses.json merl-an_bench/commands.json
[
  {
    "sample_id": "562",
    "responses": [
      {
        "class": "return",
        "notifications": [],
        "timing": {
          "clock": 460046,
          "cpu": 459871,
          "query": 459833,
          "pp": 0,
          "reader": 1,
          "ppx": 36,
          "typer": 1,
          "error": 0
        }
      }
    ],
    "merlin_id": 0,
    "cmd": "ocamlmerlin server type-enclosing -position '71:8' -filename /projects/irmin/src/irmin-git/unix/xgit.ml < /projects/irmin/src/irmin-git/unix/xgit.ml"
  },
  {
    "sample_id": "23",
    "responses": [
      {
        "class": "return",
        "notifications": [],
        "timing": {
          "clock": 344728,
          "cpu": 344599,
          "query": 344563,
          "pp": 0,
          "reader": 0,
          "ppx": 35,
          "typer": 1,
          "error": 0
        }
      }
    ],
    "merlin_id": 0,
    "cmd": "ocamlmerlin server type-enclosing -position '28:34' -filename /projects/irmin/bench/irmin-pack/main.ml < /projects/irmin/bench/irmin-pack/main.ml"
  },
  {
    "sample_id": "495",
    "responses": [
      {
        "class": "return",
        "notifications": [],
        "timing": {
          "clock": 178729,
          "cpu": 178662,
          "query": 178621,
          "pp": 0,
          "reader": 7,
          "ppx": 32,
          "typer": 2,
          "error": 0
        }
      }
    ],
    "merlin_id": 0,
    "cmd": "ocamlmerlin server type-enclosing -position '256:41' -filename /projects/irmin/src/irmin-git/irmin_git.ml < /projects/irmin/src/irmin-git/irmin_git.ml"
  },
  {
    "sample_id": "2820",
    "responses": [
      {
        "class": "return",
        "notifications": [],
        "timing": {
          "clock": 159372,
          "cpu": 159311,
          "query": 159267,
          "pp": 0,
          "reader": 3,
          "ppx": 39,
          "typer": 2,
          "error": 0
        }
      }
    ],
    "merlin_id": 0,
    "cmd": "ocamlmerlin server type-enclosing -position '260:62' -filename /projects/irmin/test/irmin-pack/test_hashes.ml < /projects/irmin/test/irmin-pack/test_hashes.ml"
  },
  {
    "sample_id": "2862",
    "responses": [
      {
        "class": "return",
        "notifications": [],
        "timing": {
          "clock": 156917,
          "cpu": 156855,
          "query": 156797,
          "pp": 0,
          "reader": 9,
          "ppx": 47,
          "typer": 2,
          "error": 0
        }
      }
    ],
    "merlin_id": 0,
    "cmd": "ocamlmerlin server type-enclosing -position '503:48' -filename /projects/irmin/test/irmin-pack/test_lower.ml < /projects/irmin/test/irmin-pack/test_lower.ml"
  },
  {
    "sample_id": "2982",
    "responses": [
      {
        "class": "return",
        "notifications": [],
        "timing": {
          "clock": 155576,
          "cpu": 155517,
          "query": 155516,
          "pp": 0,
          "reader": 0,
          "ppx": 0,
          "typer": 1,
          "error": 0
        }
      }
    ],
    "merlin_id": 0,
    "cmd": "ocamlmerlin server type-enclosing -position '28:40' -filename /projects/irmin/test/irmin-tezos/irmin_fsck.ml < /projects/irmin/test/irmin-tezos/irmin_fsck.ml"
  },
  {
    "sample_id": "902",
    "responses": [
      {
        "class": "return",
        "notifications": [],
        "timing": {
          "clock": 28800,
          "cpu": 28790,
          "query": 28759,
          "pp": 0,
          "reader": 2,
          "ppx": 28,
          "typer": 1,
          "error": 0
        }
      }
    ],
    "merlin_id": 0,
    "cmd": "ocamlmerlin server type-enclosing -position '89:14' -filename /projects/irmin/src/irmin-pack/mem/irmin_pack_mem.ml < /projects/irmin/src/irmin-pack/mem/irmin_pack_mem.ml"
  },
  {
    "sample_id": "1413",
    "responses": [
      {
        "class": "return",
        "notifications": [],
        "timing": {
          "clock": 15948,
          "cpu": 15941,
          "query": 15873,
          "pp": 0,
          "reader": 17,
          "ppx": 49,
          "typer": 2,
          "error": 0
        }
      }
    ],
    "merlin_id": 0,
    "cmd": "ocamlmerlin server type-enclosing -position '701:44' -filename /projects/irmin/src/irmin-pack/unix/store.ml < /projects/irmin/src/irmin-pack/unix/store.ml"
  },
  {
    "sample_id": "725",
    "responses": [
      {
        "class": "return",
        "notifications": [],
        "timing": {
          "clock": 6121,
          "cpu": 6118,
          "query": 6078,
          "pp": 0,
          "reader": 2,
          "ppx": 37,
          "typer": 1,
          "error": 0
        }
      }
    ],
    "merlin_id": 0,
    "cmd": "ocamlmerlin server type-enclosing -position '142:25' -filename /projects/irmin/src/irmin-pack-tools/tezos_explorer/files.ml < /projects/irmin/src/irmin-pack-tools/tezos_explorer/files.ml"
  },
  {
    "sample_id": "678",
    "responses": [
      {
        "class": "return",
        "notifications": [],
        "timing": {
          "clock": 4912,
          "cpu": 4910,
          "query": 3,
          "pp": 0,
          "reader": 13,
          "ppx": 0,
          "typer": 4894,
          "error": 0
        }
      }
    ],
    "merlin_id": 0,
    "cmd": " ocamlmerlin server locate -look-for ml -position '292:10' -index 0 -filename /projects/irmin/src/irmin-mirage/git/irmin_mirage_git.ml < /projects/irmin/src/irmin-mirage/git/irmin_mirage_git.ml"
  }
]
```

It uses a pretty hacky jq oneliner, but it does it's job. Doing it in a nice and clean way would require much more effort

**Comment by 3Rafal** [comment]
Interestingly, the top longest query run separately on my machine executes super quickly:
```
rafal@memepad:~/Projects/Tarides/irmin$ time ocamlmerlin server type-enclosing -position '71:8' -filename src/irmin-git/unix/xgit.ml < src/irmin-git/unix/xgit.ml
{"class":"return","value":[{"start":{"line":42,"col":13},"end":{"line":81,"col":3},"type":"sig end","tail":"no"},{"start":{"line":42,"col":0},"end":{"line":81,"col":3},"type":"sig end","tail":"no"}],"notifications":[],"timing":{"clock":75,"cpu":45,"query":1,"pp":0,"reader":3,"ppx":38,"typer":4,"error":0}}

real	0m0,078s
user	0m0,002s
sys	0m0,001s
```

**Comment by 3Rafal** [comment]
I've been able to reproduce a long running query with
```
ocamlmerlin server type-enclosing -position '28:34' -filename bench/irmin-pack/main.ml < bench/irmin-pack/main.ml
```

I've redirected the output to file and it consisted of 1 line with 4427193 characters and weighed almost 5 MB.

**Comment by voodoos** [comment]
> I've been able to reproduce a long running query with
> 
> ```
> ocamlmerlin server type-enclosing -position '28:34' -filename bench/irmin-pack/main.ml < bench/irmin-pack/main.ml
> ```
> 
> I've redirected the output to file and it consisted of 1 line with 4427193 characters and weighed almost 5 MB. See it attached here (warning: it can kill your terminal if you print it): [looooong.txt](https://github.com/ocaml/merlin/files/11946329/looooong.txt)

That's a good find. Is it slow even if you pipe to a file ?

**Comment by 3Rafal** [comment]
Yes, it's slow when redirecting to a file

**Comment by voodoos** [comment]
Ok,the issue is that the very long type is return for every growing enclosing. 
But in fact, the emacs and lsp modes should only ask for one enclosing at a time.
Could you try with `-index 0` ?

`ocamlmerlin server type-enclosing -position '28:34' -filename bench/irmin-pack/main.ml < bench/irmin-pack/main.ml`

I think `merl-an` should use that command for it's benchmarks since the editor modes do.

**Comment by 3Rafal** [comment]
Now it's much better :)
```
rafal@memepad:~/Projects/Tarides/irmin$ ocamlmerlin server type-enclosing -position '28:34' -index 0 -filename bench/irmin-pack/main.ml < bench/irmin-pack/main.ml
{"class":"return","value":[{"start":{"line":28,"col":17},"end":{"line":28,"col":35},"type":"functor (Config : Irmin_pack.Conf.S) -> Irmin_pack_unix__Store_intf.KV","tail":"no"},{"start":{"line":28,"col":17},"end":{"line":28,"col":35},"type":1,"tail":"no"},{"start":{"line":28,"col":17},"end":{"line":28,"col":44},"type":2,"tail":"no"},{"start":{"line":28,"col":2},"end":{"line":28,"col":44},"type":3,"tail":"no"},{"start":{"line":27,"col":12},"end":{"line":30,"col":3},"type":4,"tail":"no"},{"start":{"line":27,"col":0},"end":{"line":30,"col":3},"type":5,"tail":"no"}],"notifications":[],"timing":{"clock":41,"cpu":17,"query":2,"pp":0,"reader":0,"ppx":13,"typer":1,"error":0}}
```

**Review by voodoos — APPROVED** [review]
Looks good to me !

Do you know if the previous run will be erased when merging ? It's a shame to have the history of weird `type-enclosing` results.

**Comment by 3Rafal** [comment]
> Do you know if the previous run will be erased when merging ? It's a shame to have the history of weird type-enclosing results.

I don't think so. It would be a pretty inconvenient design tbh...
I guess that benchmarks are for given branch, so after we merge it it will be run for main branch and then it will be saved

**Comment by voodoos** [comment]
Let's give it a try then :-)

**Pull request merged**